KAFKA-7466: Add IncrementalAlterConfigs API (KIP-339)

omkreddy opened the pull request:
- https://cwiki.apache.org/confluence/display/KAFKA/KIP-339%3A+Create+a+new+IncrementalAlterConfigs+API

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

omkreddy committed:
KAFKA-7466: Add IncrementalAlterConfigs API (KIP-339)

omkreddy commented:
@cmccabe Please take a look when you get a chance. I will be adding few more tests.  
There are few minor deviations from the KIP . will update KIP and mailing after the review.

omkreddy commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -388,6 +388,47 @@ public AlterConfigsResult alterConfigs(Map<ConfigResource, Config> configs) {
+     * @param configs         The resources with their configs
+     * @return                The IncrementalAlterConfigsResult
+     */
+    public IncrementalAlterConfigsResult alterConfigs(Collection<AlterConfigOp> configs) {
```
Can we change the API to  ```alterConfigs(Map<ConfigResource, Collection<AlterConfigOp>> configs.```. 
This will be similar to existing ```alterConfigs(Map<ConfigResource, Config> configs)``` API.

ijuma replied:
I think that's what I suggested in the KIP and Colin eventually agreed.

omkreddy replied:
Updated the API. Will update the KIP after the review.

viktorsomogyi commented on clients/src/main/java/org/apache/kafka/clients/admin/AlterConfigOp.java:
```diff
@@ -0,0 +1,98 @@
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class AlterConfigOp {
```
Do we need `@InterfaceStability.Evolving` here?

stanislavkozlovski replied:
Considering the old `AlterConfigsOptions.java` still has that annotation (it's from June 2017), we could probably have this here.
We should consider removing it from the old one I think.

omkreddy committed:
Update AuthorizerIntegrationTest

omkreddy committed:
refactor incrementalAlterConfig API

omkreddy commented:
retest this please

omkreddy committed:
Update few DynamicBrokerReconfigurationTest tests to use incrementalAlterConfig API

omkreddy commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -307,7 +309,7 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
       Paths.get(sslProperties1.getProperty(SSL_TRUSTSTORE_LOCATION_CONFIG)),
       StandardCopyOption.REPLACE_EXISTING)
-    TestUtils.alterConfigs(servers, adminClients.head, oldTruststoreProps, perBrokerConfig = true).all.get()
+    TestUtils.incrementalAlterConfigs(servers, adminClients.head, oldTruststoreProps, perBrokerConfig = true).all.get()
```
@rajinisivaram i have updated few Dynamic Reconfig test cases to use incremental alter config API. Please take a look when you get a chance. Let me know, if we need to add/update any specific test cases.

stanislavkozlovski commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -388,6 +388,48 @@ public AlterConfigsResult alterConfigs(Map<ConfigResource, Config> configs) {
+    /**
+     * Incrementally update the configuration for the specified resources.
+     *
+     * Updates are not transactional so they may succeed for some resources while fail for others. The configs for
```
Should we mention this in the KIP? It was not specified there - even though this is the same behavior as the old AlterConfigs, in the context of the KIP this non-transactional functionality isn't obvious I fear

omkreddy replied:
will update the KIP

amareshb replied:
@omkreddy  KIP mentions this is `transactional functionality` but the code here says the opposite. Can you please clarify ? 

From the KIP in proposed changes section  - 
`Similar to existing alterConfigs API, we will to keep the "transactionality" of updating several configs  for the same ConfigResource at once. We guarantee that we never do a partial update of a collection of configs
for a ConfigResource from a single request. `

stanislavkozlovski commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -388,6 +388,48 @@ public AlterConfigsResult alterConfigs(Map<ConfigResource, Config> configs) {
+     *   <li>{@link org.apache.kafka.common.errors.TopicAuthorizationException}
+     *   if the authenticated user didn't have alter access to the Topic.</li>
+     *   <li>{@link org.apache.kafka.common.errors.InvalidRequestException}
+     *   if the request details are invalid.</li>
```
nit: maybe we could add an example here like "e.g a configuration key was specified more than once"

stanislavkozlovski commented on clients/src/main/java/org/apache/kafka/clients/admin/AlterConfigOp.java:
```diff
@@ -0,0 +1,96 @@
+    public enum OpType {
+        SET((byte) 0), DELETE((byte) 1), APPEND((byte) 2), SUBSTRACT((byte) 3);
+
+        private static final Map<Byte, OpType> OPTYPES = Collections.unmodifiableMap(
```
nit: Should this be `OP_TYPES`?

stanislavkozlovski commented on clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java:
```diff
@@ -1078,6 +1078,10 @@ public ConfigKey(String name, Type type, Object defaultValue, Validator validato
             return !NO_DEFAULT_VALUE.equals(this.defaultValue);
         }
+
+        public Type getType() {
```
nit: `type()`? Was that our convention across Kafka?

stanislavkozlovski commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -437,6 +492,37 @@ class AdminManager(val config: KafkaConfig,
+      alterConfigOp.opType() match {
+        case OpType.SET => configProps.setProperty(alterConfigOp.configEntry().name(), alterConfigOp.configEntry().value())
+        case OpType.DELETE => configProps.remove(alterConfigOp.configEntry() .name())
+        case OpType.APPEND =>  {
```
nit: two spaces in between `=>  {`

stanislavkozlovski commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -1378,4 +1387,30 @@ private ElectPreferredLeadersResponse createElectPreferredLeadersResponse() {
+        IncrementalAlterConfigsRequestData data = new IncrementalAlterConfigsRequestData();
+        AlterableConfig alterableConfig = new AlterableConfig()
+                .setName("retention.ms")
+                .setConfigOperation((byte) 0)
```
nit: Is there any reason to not use what you normally would when building the request - e.g `AlterConfigOp.OpType.APPEND.id()` ?

omkreddy replied:
Thus is just to verify request serialization . Also we need to update import-control .xml to use AlterConfigOp here.

stanislavkozlovski replied:
Makes sense

stanislavkozlovski commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -364,59 +367,111 @@ class AdminManager(val config: KafkaConfig,
+      try {
+        //throw InvalidRequestException if any duplicate keys
+        alterConfigOps.groupBy(config => config.configEntry().name()).mapValues(_.size).values.foreach(count =>
+            if (count > 1) throw new InvalidRequestException(s"Duplicate config keys provided for resource $resource"))
```
Will this abort all the configurations for the resource?
The KIP says: 
> If a configuration key was specified more than once in the request, INVALID_REQUEST will be returned for all occurrences of the configuration key.

I think that implies that only the config key will be denied, not all other keys related to that resource. I see that the original `AlterConfigs` appears to work in the same way though

omkreddy replied:
We may need to update the wordings of the KIP.
If there is any validation issue with configs, we fail the alter operation for that resource.  We have single status of the alter operation on each resource. If any validation fails, we fail the complete operation. We have same behaviour for existing AlterConfigs.

stanislavkozlovski replied:
I think that we should **definitely** mention this in the KIP

cmccabe replied:
The different operations should all be independent and fail or succeed independently.

stanislavkozlovski commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -364,59 +367,111 @@ class AdminManager(val config: KafkaConfig,
+        alterConfigOps.groupBy(config => config.configEntry().name()).mapValues(_.size).values.foreach(count =>
+            if (count > 1) throw new InvalidRequestException(s"Duplicate config keys provided for resource $resource"))
+
+        val configEntriesMap = alterConfigOps.map(entry => (entry.configEntry().name(), entry.configEntry().value())).toMap
```
nit: In scala, do we use parentheses for simple getters? I remember we don't but I very well may be wrong

omkreddy replied:
we are not strictly following this. but i will update if any concerns.
cc @ijuma

stanislavkozlovski commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -437,6 +492,37 @@ class AdminManager(val config: KafkaConfig,
+          val newValue =  oldValue ::: alterConfigOp.configEntry().value().split(",").toList
+          configProps.setProperty(alterConfigOp.configEntry().name(), newValue.mkString(","))
+        }
+        case OpType.SUBSTRACT => {
```
nit: typo - `SUBSTRACT` => `SUBTRACT`

stanislavkozlovski commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -437,6 +492,37 @@ class AdminManager(val config: KafkaConfig,
+          if (!listType(alterConfigOp.configEntry().name(), configKeys))
+            throw new InvalidRequestException(s"Config value substract is not allowed for config: ${alterConfigOp.configEntry().name()}")
+          val oldValue = configProps.getProperty(alterConfigOp.configEntry().name()).split(",").toList
+          val newValue =  oldValue diff alterConfigOp.configEntry().value().split(",").toList
```
Do we have the convention to use this style of method invocation in Kafka? Asking because I haven't seen it yet and I find it a bit confusing

omkreddy replied:
i changed to normal method invocation.

stanislavkozlovski replied:
Thanks

stanislavkozlovski commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -437,6 +492,37 @@ class AdminManager(val config: KafkaConfig,
+          if (!listType(alterConfigOp.configEntry().name(), configKeys))
+            throw new InvalidRequestException(s"Config value append is not allowed for config: ${alterConfigOp.configEntry().name()}")
+          val oldValue = configProps.getProperty(alterConfigOp.configEntry().name()).split(",").toList
+          val newValue =  oldValue ::: alterConfigOp.configEntry().value().split(",").toList
```
Do we need to be handling cases where this could result in duplicate values in the list after the append?

stanislavkozlovski replied:
Ah, maybe the  `LogConfig.validate(configs)` in `adminZkClient.validateTopicConfig(topic, configProps)` will catch that

stanislavkozlovski commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -364,59 +367,111 @@ class AdminManager(val config: KafkaConfig,
         resource.`type` match {
-          case ConfigResource.Type.TOPIC =>
-            val topic = resource.name
+          case ConfigResource.Type.TOPIC => alterTopicConfigs(resource, validateOnly, configProps, configEntriesMap)
```
👍 on the refactor

stanislavkozlovski commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -437,6 +492,37 @@ class AdminManager(val config: KafkaConfig,
     }.toMap
   }
 
+  private def prepareConfigs(alterConfigOps: List[AlterConfigOp], configProps: Properties, configKeys: Map[String, ConfigKey]): Unit = {
```
nit: What do you think about a more descriptive name like `prepareIncrementalConfigs`?

stanislavkozlovski commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -1399,6 +1397,158 @@ class AdminClientIntegrationTest extends IntegrationTestHarness with Logging {
+
+    assertFutureExceptionTypeEquals(alterResult.values().get(topicResource1), classOf[InvalidRequestException],
+      Some("Invalid config value for resource"))
+  }
```
Should we also add a test case against lost updates? e.g `a=1,b=2,c=3`, incremental config change - `a=4,b=5` and we ensure that `c=3` still

stanislavkozlovski replied:
I think this can piggy-back on the `//verify config value substract` check

omkreddy replied:
added a check to verify previous updates are intact.

stanislavkozlovski commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -1399,6 +1397,158 @@ class AdminClientIntegrationTest extends IntegrationTestHarness with Logging {
+  }
+
+  @Test
+  def testInValidIncrementalAlterConfigs(): Unit = {
```
nit: capitalization `InValid` => `Invalid`

stanislavkozlovski commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -1399,6 +1397,158 @@ class AdminClientIntegrationTest extends IntegrationTestHarness with Logging {
+
+    var topicConfigEntries1 = Seq(
+      new AlterConfigOp(new ConfigEntry(LogConfig.CleanupPolicyProp, LogConfig.Delete), AlterConfigOp.OpType.APPEND),
+      new AlterConfigOp(new ConfigEntry(LogConfig.CleanupPolicyProp, LogConfig.Compact), AlterConfigOp.OpType.DELETE)
```
I think that we should also add a new, **valid** config entry here and ensure that it gets updated despite the invalid `cleanup.policy`. That's how I understand the KIP should work at least.

If that's not the intention, maybe we could add a valid config entry and assert that it doesn't get updated.

omkreddy replied:
We don't do partial update for a resource. either we apply all configs or fail the request. 
Updated the test to add a valid config entry and assert that it doesn't get updated.

cmccabe replied:
The intention in KIP-339 is to allow partial updates. Each AlterConfigOp is independent and just because one fails, doesn't mean any other one will fail (except in the case of duplicate keys as described earlier).

omkreddy replied:
@cmccabe Updated the PR. There are scenarios, where we need to update/validate related configs in single batch. In this cases, each AlterConfigOp are not  independent .
Example: while adding listener, user want may want to update listener and listener.security.protocol.map configs Without listener.security.protocol.map config, listener config update may fail. 

I have added a check to decide if we need batch config validation or not. Not sure if this is right approach. Please take a look when you get a chance.

cc @rajinisivaram

stanislavkozlovski commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -1399,6 +1397,158 @@ class AdminClientIntegrationTest extends IntegrationTestHarness with Logging {
+
+    assertFutureExceptionTypeEquals(alterResult.values().get(topicResource2), classOf[InvalidRequestException],
+      Some("Config value substract is not allowed for config"))
+  }
```
So far we've only tested against IncrementalAlterConfig semantics. What do you think about adding a quick assert about an invalid property value to ensure those get validated as well?  This can be equivalent to the 
```java
      new ConfigEntry(LogConfig.MinCleanableDirtyRatioProp, "1.1"), // this value is invalid as it's above 1.0
```
check made in `def checkInvalidAlterConfigs`

omkreddy replied:
updated the test

stanislavkozlovski commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -1433,6 +1448,8 @@ object TestUtils extends Logging {
         val cause = e.getCause()
         assertTrue("Expected an exception of type " + clazz.getName + "; got type " +
             cause.getClass().getName, clazz.isInstance(cause))
+        errorMessage.foreach(message => assertTrue(s"Wrong message ${cause.getMessage}", cause.getMessage.contains(message)))
```
nit: Could we change the error message to be more informative? e.g `Message X did not contain message Y as expected"
I also suggest we change the `errorMessage` variable name to `errorMessageSubset` (or similar) to make it more obvious that we want the message to be contained and not directly equal the message we get. 

The method name `assertFutureExceptionTypeEquals` would leave me to expect that we want everything to be 1:1 equal otherwise

omkreddy replied:
for now, i updated the message.

stanislavkozlovski reviewed (commented):
Thanks for picking up this KIP, @omkreddy!

I left a bunch of nit comments related to style. I think this is sound functionality-wise.

As a general question - I don't see us marking the old `AlterConfigs` as deprecated anywhere but the KIP does mention "We should deprecate AlterConfigs." in the motivation

omkreddy committed:
Address stan's comments

omkreddy commented:
@stanislavkozlovski Thanks for the detailed review. I have addressed most of the review comments.
alterConfigs API is deprecated in favour of new API.

omkreddy commented:
retest this please

stanislavkozlovski reviewed (approved):
Thanks for addressing my small comments. Glad we deprecated AlterConfigs. LGTM! :)

ijuma commented:
My high level comment here is that we should think about how we can use the short name for the non deprecated API. A few options:

1. For the requests, we can just rename the old ones and use `AlterConfigs` for the new ones. Since the id is what uniquely identifies the request type.
2. For AdminClient, we agreed to overload the method names. Can we do anything about the option and result classes?

omkreddy commented:
@ijuma Makes sense.

1. We can rename old requests/responses to LegacyAlterConfigsRequest/Response.
2. We can reuse existing AlterConfigsResult/AlterConfigsOptions options
3. One issue with method overload is, both methods have same type erasure.
 
'alterConfigs(Map<ConfigResource, Config>)' clashes with 'alterConfigs(Map<ConfigResource, Collection<AlterConfigOp>>)'; both methods have same erasure

'alterConfigs(Map<ConfigResource, Config>, AlterConfigsOptions)' clashes with 'alterConfigs(Map<ConfigResource, Collection<AlterConfigOp>>, AlterConfigsOptions)'; both methods have same erasure 

cc @cmccabe

cmccabe commented:
> We can rename old requests/responses to LegacyAlterConfigsRequest/Response.

I think that's OK.  It's internal, after all, and always identified by its ID.

> We can reuse existing AlterConfigsResult/AlterConfigsOptions options

Yes, this is fine.  The options match up for now.  If we later add stuff to `AlterConfigsOptions` that is not relevant to the legacy calls, we can have the legacy calls fail with an error when the relevant options are specified.

> One issue with method overload is, both methods have same type erasure.

This doesn't match up with what we agreed to in the KIP.  The KIP has:
```
public IncrementalAlterConfigsResult incrementalAlterConfigs(
    Collection<AlterConfigOp> ops,
    final IncrementalAlterConfigsOptions options);
```
The `AlterConfigOp` objects are specified as a collection, not as a map.

One important thing about the AlterConfigsResult class: I think we need to have a way to get the error for a particular AlterConfigOp.  Something like `KafkaFuture<Void> opResult(AlterConfigOp)`  The values() method is just a summary method, similar to all().  Since the same resource may have AlterConfigOps that failed, and also AlterConfigOps that succeeded.

omkreddy commented:
@cmccabe Thanks for the review,.

>One important thing about the AlterConfigsResult class: I think we need to have a way to get the error >for a particular AlterConfigOp. 

[KIP-339](https://cwiki.apache.org/confluence/display/KAFKA/KIP-339%3A+Create+a+new+IncrementalAlterConfigs+API#KIP-339:CreateanewIncrementalAlterConfigsAPI-ProposedChanges) response protocol does not define result status for each AlterConfigOp/ConfigKey. So I have implemented the API ```alterConfigs(Map<ConfigResource, Collection< AlterConfigOp >) ```,  which will return status for each resource. .  I have assumed few things for first version of the PR.  I will update the response protocol to return status for each AlterConfigOp/Config key.

cmccabe commented:
Yeah, I think it's much better to have a status for each Op.  Otherwise it could be difficult to figure out which particular configuration failed.  Some of this is probably unclear in the KIP so we should post to the KIP thread once we figure it out.  Thanks, @omkreddy !

rajinisivaram commented on checkstyle/import-control.xml:
```diff
@@ -147,6 +147,8 @@
       <allow pkg="org.apache.kafka.common.resource" />
       <allow pkg="org.apache.kafka.common.record" />
+      <allow class="org.apache.kafka.clients.admin.AlterConfigOp" />
+      <allow class="org.apache.kafka.clients.admin.ConfigEntry" />
```
We shouldn't have references from `common` to `clients`. We fixed one of these earlier in https://issues.apache.org/jira/browse/KAFKA-7498.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/AlterConfigOp.java:
```diff
@@ -0,0 +1,106 @@
+public class AlterConfigOp {
+
+    public enum OpType {
+        SET((byte) 0), DELETE((byte) 1), APPEND((byte) 2), SUBTRACT((byte) 3);
```
ADD/REMOVE would be better than APPEND/SUBTRACT?

cmccabe replied:
I think ADD might be too easily confused with PUT.  APPEND is going to be mostly about appending to string-based keys, so the existing name makes sense, I think.  What do you think?

rajinisivaram reviewed (commented):
@omkreddy Thanks for the PR, left a couple of comments.

I am not sure the API proposed in the KIP fits with the type of updates we support. The old API with Map<ConfigResource, Config> fits better and we need to find a way to do something similar while still retaining the old one. As you already mentioned we can't just overload the method unless we also change the method in some way.

We should have a collection of updates for each `ConfigResource` with results returned per-ConfigResource since that is our unit of atomicity. We guarantee that we never do a partial update of a collection of configs for a `ConfigResource`  from a single request and hence we should only have one input with a collection of updates and one result for each ConfigResource, making the model obvious in the API and docs.

Will add a comment on the KIP thread for discussion.

cmccabe commented:
Good point, @rajinisivaram.  I posted a proposal on the mailing list which I think does keep the transactionality:

> How about an input like Map<ConfigResource, List<ConfigOps>> and a result like: Map<ConfigResource, ApiError>?

omkreddy committed:
Use existing AlterConfigsResult, AlterConfigsOptions API classes instread of new IncrementalAlterConfigsResult, IncrementalAlterConfigsOptions.

omkreddy commented:
@rajinisivaram @cmccabe @ijuma  Thanks for reviews and suggestions. Updated the code to reflect the changes discussed in mailing list. 

As discussed in one of the earlier  [comments](https://github.com/apache/kafka/pull/6247#issuecomment-470623744  ),  I have used AlterConfigsResult, AlterConfigsOptions API classes instead of new IncrementalAlterConfigsResult, IncrementalAlterConfigsOptions. Updated the KIP.

One pending item is,  shorter name for the new API method.. If we want to maintain same as old API (alterAlterConfigs), we need to handle same type erasure error.

- One option is to use `Supplier` as parameter in method signature.
  `public AlterConfigsResult                                                                                                                       
  alterAlterConfigs(Supplier<Map<ConfigResource, Collection<AlterConfigOp>>> configs)`

- Another option is to use different name `updateConfigs`, `modifyConfigs`?

cmccabe commented:
> One pending item is, shorter name for the new API method.. If we want to maintain same as old API (alterAlterConfigs), we need to handle same type erasure error.

Good point.  How about putting `AlterConfigsOptions` first in the arguments list, before the `Map`?  That seems nicer than adding `Supplier`.

Then we can add `@deprecated` to the old function and put a big warning there about why you should NOT use it. :)

I think we considered and rejected the idea of a different name like `modifyConfigs`, etc. etc.  There were a lot of discussions about this.

omkreddy commented:
@cmccabe Thanks for review.  Yes, We can put  `AlterConfigsOptions` first in the arguments list, before the Map. But, we can not have convenience method for  `AdminClient#alterConfigs(AlterConfigsOptions, Map)` with default options.
`public AlterConfigsResult alterConfigs(Map<> configs)`

If that is OK, then i am fine with this option.

cmccabe commented:
> @cmccabe Thanks for review. Yes, We can put AlterConfigsOptions first in the arguments list, before the Map. But, we can not have convenience method for AdminClient#alterConfigs(AlterConfigsOptions, Map) with default options.  public AlterConfigsResult alterConfigs(Map<> configs)

Yeah, that is annoying.  I didn't think of that. :(

Let's call the function `incrementalAlterConfigs` for now, and try to get in this PR, and then continue the discussion.

omkreddy commented:
retest this please

cmccabe commented:
Failing test is unrelated.
LGTM

cmccabe commented:
Thanks, @omkreddy!  I have committed this.  We can do any follow-up changes in separate PRs.  Nice to have the KIP-339 API in.

omkreddy commented:
@stanislavkozlovski @cmccabe @ijuma  @rajinisivaram @viktorsomogyi 
Thanks for all the reviews!.
